exercises/joseph_ramsay/ex02_oo_basics/src/shapeSorter.cpp

```diff
@@ -0,0 +1,70 @@
+#include "shapeSorter.h"
+#include <algorithm>
+
+ShapeSorter::ShapeSorter(std::vector<Shape *> shapes) : m_shapes(shapes){};
+
+void ShapeSorter::setShapes(const std::vector<Shape *> shapes) {
+  m_shapes = shapes;
+}
+
+/** General-purpose function for printing every shape which satisfies a given
+ * condition
+ *
+ * @param pred A lambda for which shapes should be printed
+ *
+ */
+void ShapeSorter::printOnCondition(
+    const std::function<bool(Shape *shape)> &pred) const {
+  // Iterate over shapes, print those for which pred(shape) evaluates to true
+  for_each(m_shapes.begin(), m_shapes.end(), [=](Shape *shape) {
+    if (pred(shape)) {
+      std::cout << shape->toString() << " ";
+    }
+  });
+  std::cout << "\n";
+}
+
+/** Print all shapes of a given type
+ *
+ * @param type The type of shapes which should be printed
+ *
+ */
+void ShapeSorter::shapesOfType(const ShapeType &type) const {
+  printOnCondition([=](Shape *shape) { return shape->getType() == type; });
```

Comment:
is `&` for the lambda capture enough (not that it matters too much in this case)?